File: backend/src/utils/latencyTracker.js
const fs = require('fs').promises;
const path = require('path');

class LatencyTracker {
  constructor() {
    this.recentLatencies = [];
    this.maxRecentEntries = 10;
    this.logFile = path.join(__dirname, '../../logs/latency.log');
    this.ensureLogDirectory();
  }

  async ensureLogDirectory() {
    const logDir = path.dirname(this.logFile);
    try {
      await fs.mkdir(logDir, { recursive: true });
    } catch (error) {
      console.error('Failed to create logs directory:', error);
    }
  }

  /**
   * Record a new latency measurement
   */
  recordLatency(endpoint, method, latencyMs, status = 'success') {
    const timestamp = new Date().toISOString();
    const entry = {
      timestamp,
      endpoint,
      method,
      latency_ms: latencyMs,
      status
    };

    // Add to recent latencies (keep only last 10)
    this.recentLatencies.push(entry);
    if (this.recentLatencies.length > this.maxRecentEntries) {
      this.recentLatencies.shift();
    }

    // Log to file asynchronously
    this.logToFile(entry).catch(error => {
      console.error('Failed to log latency:', error);
    });
  }

  /**
   * Get average latency of recent requests
   */
  getAverageLatency() {
    if (this.recentLatencies.length === 0) return 0;
    
    const successfulEntries = this.recentLatencies.filter(entry => entry.status === 'success');
    if (successfulEntries.length === 0) return 0;
    
    const totalLatency = successfulEntries.reduce((sum, entry) => sum + entry.latency_ms, 0);
    return Math.round(totalLatency / successfulEntries.length);
  }

  /**
   * Get recent latency statistics
   */
  getLatencyStats() {
    const successfulEntries = this.recentLatencies.filter(entry => entry.status === 'success');
    
    if (successfulEntries.length === 0) {
      return {
        count: 0,
        average: 0,
        min: 0,
        max: 0,
        recent_entries: 0
      };
    }

    const latencies = successfulEntries.map(entry => entry.latency_ms);
    
    return {
      count: successfulEntries.length,
      average: Math.round(latencies.reduce((a, b) => a + b, 0) / latencies.length),
      min: Math.min(...latencies),
      max: Math.max(...latencies),
      recent_entries: this.recentLatencies.length
    };
  }

  /**
   * Log entry to file
   */
  async logToFile(entry) {
    try {
      const logLine = `${entry.timestamp} | ${entry.method} ${entry.endpoint} | ${entry.latency_ms}ms | ${entry.status}\n`;
      await fs.appendFile(this.logFile, logLine);
    } catch (error) {
      // Silently fail to avoid infinite loops
    }
  }

  /**
   * Get recent log entries
   */
  async getRecentLogs(lines = 50) {
    try {
      const data = await fs.readFile(this.logFile, 'utf8');
      const logLines = data.trim().split('\n');
      return logLines.slice(-lines);
    } catch (error) {
      return [];
    }
  }

  /**
   * Clear all logs
   */
  async clearLogs() {
    try {
      await fs.writeFile(this.logFile, '');
      this.recentLatencies = [];
    } catch (error) {
      console.error('Failed to clear logs:', error);
    }
  }
}

// Singleton instance
const latencyTracker = new LatencyTracker();

module.exports = latencyTracker;

File: fastapi-backend/app/utils/latency_tracker.py
import time
import asyncio
from typing import List, Dict, Any
from collections import deque
import json
from datetime import datetime
import aiofiles
import os

class FastAPILatencyTracker:
    def __init__(self, max_recent_entries: int = 10):
        self.recent_latencies = deque(maxlen=max_recent_entries)
        self.log_file = "logs/fastapi_latency.log"
        self.ensure_log_directory()
    
    def ensure_log_directory(self):
        """Ensure logs directory exists"""
        os.makedirs(os.path.dirname(self.log_file), exist_ok=True)
    
    async def record_latency(self, endpoint: str, method: str, latency_ms: float, status: str = "success", details: Dict = None):
        """Record a new latency measurement"""
        timestamp = datetime.utcnow().isoformat()
        entry = {
            "timestamp": timestamp,
            "endpoint": endpoint,
            "method": method,
            "latency_ms": latency_ms,
            "status": status,
            "details": details or {}
        }
        
        # Add to recent latencies
        self.recent_latencies.append(entry)
        
        # Log to file asynchronously
        try:
            await self._log_to_file(entry)
        except Exception as e:
            print(f"Failed to log latency: {e}")
    
    def get_average_latency(self) -> float:
        """Get average latency of recent successful requests"""
        if not self.recent_latencies:
            return 0.0
        
        successful_entries = [entry for entry in self.recent_latencies if entry["status"] == "success"]
        if not successful_entries:
            return 0.0
        
        total_latency = sum(entry["latency_ms"] for entry in successful_entries)
        return round(total_latency / len(successful_entries), 1)
    
    def get_latency_stats(self) -> Dict[str, Any]:
        """Get comprehensive latency statistics"""
        if not self.recent_latencies:
            return {
                "count": 0,
                "average": 0.0,
                "min": 0.0,
                "max": 0.0,
                "recent_entries": 0
            }
        
        successful_entries = [entry for entry in self.recent_latencies if entry["status"] == "success"]
        if not successful_entries:
            return {
                "count": 0,
                "average": 0.0,
                "min": 0.0,
                "max": 0.0,
                "recent_entries": len(self.recent_latencies)
            }
        
        latencies = [entry["latency_ms"] for entry in successful_entries]
        
        return {
            "count": len(successful_entries),
            "average": round(sum(latencies) / len(latencies), 1),
            "min": round(min(latencies), 1),
            "max": round(max(latencies), 1),
            "recent_entries": len(self.recent_latencies)
        }
    
    async def _log_to_file(self, entry: Dict):
        """Log entry to file"""
        try:
            log_line = f"{entry['timestamp']} | {entry['method']} {entry['endpoint']} | {entry['latency_ms']:.1f}ms | {entry['status']}\n"
            async with aiofiles.open(self.log_file, mode='a') as f:
                await f.write(log_line)
        except Exception:
            # Silently fail to avoid infinite loops
            pass
    
    async def get_recent_logs(self, lines: int = 50) -> List[str]:
        """Get recent log entries"""
        try:
            async with aiofiles.open(self.log_file, mode='r') as f:
                content = await f.read()
                log_lines = content.strip().split('\n')
                return log_lines[-lines:] if log_lines else []
        except Exception:
            return []
    
    async def clear_logs(self):
        """Clear all logs"""
        try:
            async with aiofiles.open(self.log_file, mode='w') as f:
                await f.write('')
            self.recent_latencies.clear()
        except Exception as e:
            print(f"Failed to clear logs: {e}")

# Global instance
latency_tracker = FastAPILatencyTracker()